plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -388,56 +525,245 @@ fu_rts54hub_rtd21xx_device_write_firmware (FuDevice *device,
 
 	/* open device */
 	locker = fu_device_locker_new (self, error);
-	if (locker == NULL)
+	if (locker == NULL) {
+		return fu_rts54hub_rtd21xx_device_exit_background (self, error);
+		return FALSE;
+	}
+
+	/* simple image */
+	fw = fu_firmware_get_bytes (firmware, error);
+	if (fw == NULL) {
+		fu_rts54hub_rtd21xx_device_exit_background (self, error);
```

Comment:
we can't just add these in here -- it's all very chaotic. Can you describe what foreground and background modes actually are? Could we do the same in attach and detach and avoid all these sprinkled calls?

Reply:
Background mode means that the device can work normally in the process of fw update. After burning, the device cannot be restarted.
Foreground mode means the device can't work  in the process of fw update, after FW Update it will restart. If  add exit process  to attach, you need to judge whether FW update has been performed or not, If there is a burn, you need to send resetcommand. If you do not perform burn, you need to send the command to exit FW update mode.

Reply:
If that's the case - this is basically the meaning of [FWUPD-DEVICE-FLAG-USABLE-DURING-UPDATE](https://fwupd.github.io/fwupd-fwupd-enums.html#FWUPD-DEVICE-FLAG-USABLE-DURING-UPDATE:CAPS).  Just drop the private member and set/use that flag in the plugin.